[Shell] Update logic for iOS large title display in ShellItemRenderer

kubaflo opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

## Background and Context

This PR addresses issue #33037 where iOS large titles in Shell don't transition properly when scrolling. This is the third attempt to fix the issue:

1. **PR #33039** (merged Dec 8, 2025): Changed the condition from `PrefersLargeTitles = largeTitleDisplayMode == Always` to `PrefersLargeTitles = largeTitleDisplayMode != Never`
2. **PR #33230** (merged Dec 19, 2025): Reverted #33039 because it caused "a lot of failing tests" - large titles started appearing unintentionally in many scenarios, breaking UI test snapshots
3. **PR #33246** (this PR): Re-attempts the fix with an important addition: opt-in check

## Root Cause

Shell doesn't have a `PrefersLargeTitles` property at the Shell level. The previous implementation would update large title preferences whenever any page was displayed, regardless of whether the developer explicitly requested large title behavior. 

**Problem with PR #33039:** Changing the condition to `largeTitleDisplayMode != Never` meant that:
- `LargeTitleDisplayMode.Always` → PrefersLargeTitles = true ✅
- `LargeTitleDisplayMode.Automatic` → PrefersLargeTitles = true ⚠️ (unintended for pages that didn't set the property)
- `LargeTitleDisplayMode.Never` → PrefersLargeTitles = false ✅

Since the default value is `Automatic`, pages that never set `LargeTitleDisplay` would suddenly get large titles, breaking many UI tests.

## Description of Change

This PR implements an **opt-in model** for large title display in Shell:

### 1. Added Opt-In Check
Before updating large title preferences, the code now checks if the `LargeTitleDisplay` property is explicitly set on the Page:

```csharp
if (!page.IsSet(PlatformConfiguration.iOSSpecific.Page.LargeTitleDisplayProperty))
{
    return; // Don't update if property not explicitly set
}

rmarinho commented:
## 📋 PR Finalization Review
<!-- PR-FINALIZE-COMMENT -->

<details>

<summary><b>Title: ✅ Good</b></summary>

<br>

**Current:** `[Shell] Update logic for iOS large title display in ShellItemRenderer`
</details>

<details>

<summary><b>Description: ✅ Good</b></summary>

<br>

Description needs updates. See details below.
### ✨ Suggested PR Description

# Recommended PR Description for #33246

```markdown
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

## Background and Context

This PR addresses issue #33037 where iOS large titles in Shell don't transition properly when scrolling. This is the third attempt to fix the issue:

1. **PR #33039** (merged Dec 8, 2025): Changed the condition from `PrefersLargeTitles = largeTitleDisplayMode == Always` to `PrefersLargeTitles = largeTitleDisplayMode != Never`
2. **PR #33230** (merged Dec 19, 2025): Reverted #33039 because it caused "a lot of failing tests" - large titles started appearing unintentionally in many scenarios, breaking UI test snapshots
3. **PR #33246** (this PR): Re-attempts the fix with an important addition: opt-in check

## Root Cause

Shell doesn't have a `PrefersLargeTitles` property at the Shell level. The previous implementation would update large title preferences whenever any page was displayed, regardless of whether the developer explicitly requested large title behavior. 

**Problem with PR #33039:** Changing the condition to `largeTitleDisplayMode != Never` meant that:
- `LargeTitleDisplayMode.Always` → PrefersLargeTitles = true ✅
- `LargeTitleDisplayMode.Automatic` → PrefersLargeTitles = true ⚠️ (unintended for pages that didn't set the property)
- `LargeTitleDisplayMode.Never` → PrefersLargeTitles = false ✅

Since the default value is `Automatic`, pages that never set `LargeTitleDisplay` would suddenly get large titles, breaking many UI tests.

## Description of Change

This PR implements an **opt-in model** for large title display in Shell:

### 1. Added Opt-In Check
Before updating large title preferences, the code now checks if the `LargeTitleDisplay` property is explicitly set on the Page:

```csharp
if (!page.IsSet(PlatformConfiguration.iOSSpecific.Page.LargeTitleDisplayProperty))
{
    return; // Don't update if property not explicitly set
}
```

**Effect:** Large titles will only be configured when developers explicitly opt-in by setting the `LargeTitleDisplay` property on their Page.

### 2. Improved Condition Logic
Kept the improved condition from PR #33039:

```csharp
navigationController.NavigationBar.PrefersLargeTitles = largeTitleDisplayMode != LargeTitleDisplayMode.Never;
```

**Effect:** When developers DO set the property:
- `LargeTitleDisplayMode.Always` → Enables large titles ✅
- `LargeTitleDisplayMode.Automatic` → Enables large titles (iOS decides when to show) ✅
- `LargeTitleDisplayMode.Never` → Disables large titles ✅

### Philosophy Change

**Before (PR #33039):**
- Large titles could be enabled implicitly based on `Automatic` default
- No opt-in required

**After (this PR):**
- Large titles require explicit opt-in via `LargeTitleDisplay` property
- Pages without the property set are not affected

**Future (.NET 11):** Add a `PrefersLargeTitles` property to Shell itself for better default control (see TODO comment in code).

## Technical Details

**Modified file:**
- `src/Controls/src/Core/Compatibility/Handlers/Shell/iOS/ShellItemRenderer.cs`

**Key changes:**
1. Lines 458-464: Added `IsSet()` check for opt-in behavior
2. Line 470: Updated condition from `== Always` to `!= Never`

**Platform affected:**
- iOS only (iOS 11+)

## Testing Recommendations

This PR should be tested against the following scenarios:

| Scenario | Expected Behavior |
|----------|-------------------|
| Page with no `LargeTitleDisplay` set | No large title display (early return) |
| Page with `LargeTitleDisplayMode.Always` | Large titles enabled and displayed |
| Page with `LargeTitleDisplayMode.Automatic` | Large titles enabled, iOS controls when to show |
| Page with `LargeTitleDisplayMode.Never` | Large titles disabled |
| Existing UI tests | Should pass (no unintended large titles) |

**Critical validation:** Confirm that UI tests that failed with PR #33039 now pass with this opt-in approach.

## Breaking Changes

**Potential breaking change:** Apps that relied on implicit large title behavior (without explicitly setting `LargeTitleDisplay`) will no longer get large titles. However, this is the intended behavior to align with the opt-in model.

**Migration:** Developers who want large titles should explicitly set:
```csharp
page.On<iOS>().SetLargeTitleDisplay(LargeTitleDisplayMode.Always);
// or
page.On<iOS>().SetLargeTitleDisplay(LargeTitleDisplayMode.Automatic);
```

## Issues Fixed

Fixes #33037  
Supersedes #33230 (revert)  
Replaces #33039 (reverted version)

## Platforms Tested

- [x] iOS
- [ ] Mac Catalyst (should also benefit from this fix)
```
</details>

<details>

<summary><b>Code Review: ✅ Passed</b></summary>

<br>

# Code Review Findings for PR #33246

## 🟡 Suggestions (Non-Blocking)

### 1. Fix Typo in Comment

**Location:** `ShellItemRenderer.cs:459`

**Current:**
```csharp
// if users doen't explicitly set the LargeTitleDisplay property on the Page.
```

**Should be:**
```csharp
// if users don't explicitly set the LargeTitleDisplay property on the Page.
```

**Fix:** Change "doen't" to "don't"

---

### 2. Enhance Comment Clarity

**Location:** `ShellItemRenderer.cs:458-460`

**Current:**
```csharp
// Shell doesn't have the property PrefersLargeTitles, so we should not update the large titles
// if users doen't explicitly set the LargeTitleDisplay property on the Page.
// todo net 11: Add PrefersLargeTitles to Shell and use that here.
```

**Suggested improvement:**
```csharp
// Shell doesn't have a PrefersLargeTitles property, so we should only update large title
// preferences when users explicitly opt-in by setting LargeTitleDisplay on the Page.
// This prevents unintended large title behavior when the property isn't set.
// TODO .NET 11: Add PrefersLargeTitles property to Shell for better default control.
```

**Rationale:** 
- Clarifies the "opt-in" model
- Explains the motivation (preventing unintended behavior)
- More readable TODO format

---

### 3. Add Historical Context Comment

**Location:** Above the opt-in check (line 458)

**Suggested addition:**
```csharp
// Note: This opt-in check was added after PR #33039 was reverted (PR #33230).
// Without this check, the condition change below (PrefersLargeTitles != Never) caused
// large titles to appear unintentionally in scenarios where LargeTitleDisplay wasn't
// explicitly set, breaking many UI tests.
if (!page.IsSet(PlatformConfiguration.iOSSpecific.Page.LargeTitleDisplayProperty))
{
    return;
}
```

**Rationale:** Helps future maintainers understand why this check exists and prevents accidental removal.

---

## 🔴 Critical Questions Requiring Validation

### 1. Test Coverage

**Concern:** PR #33039 was reverted due to "a lot of failing tests" and broken snapshots.

**Questions:**
1. Have the existing UI tests been validated to pass with this change?
2. Are there new tests covering the opt-in scenarios?
3. Have snapshot tests been updated if needed?

**Recommended tests to add/verify:**

```csharp
[Test]
public void LargeTitles_NotSet_DoesNotApplyLargeTitles()
{
    // Scenario: Page without LargeTitleDisplay property set
    // Expected: Large titles should NOT be configured
}

[Test]
public void LargeTitles_SetToAlways_AppliesLargeTitles()
{
    // Scenario: Page with LargeTitleDisplay = Always
    // Expected: PrefersLargeTitles = true
}

[Test]
public void LargeTitles_SetToAutomatic_AppliesLargeTitles()
{
    // Scenario: Page with LargeTitleDisplay = Automatic
    // Expected: PrefersLargeTitles = true
}

[Test]
public void LargeTitles_SetToNever_DoesNotApplyLargeTitles()
{
    // Scenario: Page with LargeTitleDisplay = Never
    // Expected: PrefersLargeTitles = false
}
```

---

### 2. Breaking Change Assessment

**Concern:** This introduces an opt-in model that changes default behavior.

**Questions:**
1. **Are there existing apps** that rely on implicit large title behavior (without explicitly setting `LargeTitleDisplay`)?
2. **Migration path:** Should the PR include migration guidance in release notes?
3. **Documentation:** Does the large title documentation need updating to emphasize the opt-in model?

**Impact analysis needed:**
- Search the codebase for usages of large titles without explicit property setting
- Check sample apps and templates

---

### 3. Default Behavior Clarity

**Concern:** What should happen when `LargeTitleDisplay` is not set?

**Current implementation:** Early return (no large title configuration)

**Questions:**
1. Is this the desired default behavior for Shell?
2. Should Shell have its own default preference?
3. Should the TODO comment about adding `PrefersLargeTitles` to Shell be prioritized?

---

## ✅ Positive Aspects

1. **Opt-in model prevents regression:** The `IsSet()` check solves the root cause of the #33039 revert
2. **Clean guard clause:** Early return pattern is clear and maintainable
3. **Platform version guard:** Maintains iOS 11+ check appropriately
4. **Forward-looking TODO:** Good note about future API improvement
5. **Minimal change:** Surgical fix that addresses the specific issue

---

## Summary

**Overall code quality:** Good implementation with minor improvements needed

**Blocking issues:** None in code itself, but **validation required** before merge:
1. ✅ Verify existing UI tests pass
2. ✅ Add/verify test coverage for opt-in scenarios
3. ✅ Assess breaking change impact on existing apps
4. ✅ Update documentation if needed

**Non-blocking improvements:**
- Fix typo
- Enhance comments
- Add historical context

**Merge recommendation:** ⚠️ **Not ready until test validation is complete and breaking change impact is assessed.**
</details>

kubaflo committed:
Add iOS Shell large-title UI tests (Issue33037)

Introduce a HostApp test page and corresponding UITest for GitHub issue 33037. The HostApp page (Issue33037.cs) sets iOS LargeTitleDisplayMode.Always, adds a scrollable list with automation IDs (TestScrollView, PageTitle, Item#) to reproduce the large-title → standard-title transition. The shared UITest (Issue33037.cs) waits for the page, captures a before-scroll screenshot, scrolls the view to trigger the title transition, and captures an after-scroll screenshot. The test file is conditionally compiled (#if TEST_FAILS_ON_ANDROID && TEST_FAILS_ON_WINDOWS) to control target platforms.

kubaflo committed:
Update logic for iOS large title display in ShellItemRenderer

Now checks if the LargeTitleDisplay property is explicitly set on the Page before updating large title preferences. Also changes the condition to set PrefersLargeTitles to true unless LargeTitleDisplayMode is Never, improving alignment with user intent.

kubaflo committed:
Add Issue33037 BeforeScroll snapshots

Add test snapshot images (Issue33037_BeforeScroll.png) for Mac and iOS (ios and ios-26) under src/Controls/tests/TestCases.*. These images are used by UI tests to validate the UI state before scrolling for issue #33037.

kubaflo committed:
Add Issue33037 AfterScroll snapshots

Add baseline snapshot images for Issue33037 AfterScroll to the test suites: Mac and iOS (ios and ios-26). These images update the visual test baselines to reflect the post-scroll state used by TestCases.
